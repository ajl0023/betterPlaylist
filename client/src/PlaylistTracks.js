import React from "react";
import { ReactComponent as CheckIcon } from "./images/check.svg";
import { ReactComponent as CheckIconPlaceH } from "./images/checkIconPlaceH.svg";
import style from "./styles/playlisttracks.module.scss";
const PlaylistTrack = (props) => {
  let track = props.playlistSet;
  const handleCheckClick = (trackid, playlistid, uri) => {
    props.handleCheckSelected(
      trackid,
      playlistid,
      uri,
      props.changed,
      props.track.position || props.index
    );
  };
  return (
    <div className={style["item-container"]}>
      <div className={style["check-count-container"]}>
        <div className={style["check-container"]}>
          <CheckIconPlaceH
            style={{ display: props.checkAll ? "flex" : "none" }}
            className={style["check-icon-place"]}
          />
          <CheckIcon
            data-testid={props.track.track.uid}
            onClick={(e) => {
              handleCheckClick(
                props.track._id || props.track.track.uid,
                props.playlist.id,
                props.track.track.uri
              );
            }}
            style={{
              display: props.isSelected ? "flex" : "none",
              fill: props.isSelected ? "#1db954" : "white",
            }}
            className={
              style[
                props.playlist.owner.display_name === "Spotify"
                  ? "check-icon"
                  : "check-icon"
              ]
            }
          />
        </div>
        <div className={style["track-count-container"]}>
          <span className={style[props.checkAll ? "inactive" : "track-count"]}>
            {props.index + 1}
          </span>
        </div>
      </div>
      <div className={style["item-title-container"]}>
        <div className={style["album-image-container"]}>
          <img
            className={style["album-image"]}
            src={
              props.imageSelection && props.playlistSet
                ? props.track.track.album.images[0] &&
                  props.track.track.album.images[0].url
                : props.playlist.images[0] && props.playlist.images[0].url
            }
            alt=""
          />
        </div>
        <div className={style["title-container"]}>
          <div className={style["track-title-container"]}>
            <p className={style["track-title"]}>{props.track.track.name}</p>
          </div>
          <span className={style["track-artist"]}>
            {props.track.track.artists.map((artist, i) => {
              return (
                <span key={artist.id}>
                  <a>
                    <span className={style["artist-name"]}>{artist.name}</span>
                  </a>
                  {props.track.track.artists.length - 1 !== i && ", "}
                </span>
              );
            })}
          </span>
        </div>
      </div>
      <div className={style["track-album-container"]}>
        <p className={style["track-album"]}>{props.track.track.album.name}</p>
      </div>
      <div className={style["track-playlist-container"]}>
        <p className={style["track-playlist"]}>{props.playlist.name}</p>
      </div>
    </div>
  );
};
export default React.memo(PlaylistTrack, (prev, next) => {
  if (
    prev.checkAll !== next.checkAll ||
    prev.lastUpdated !== next.lastUpdated
  ) {
    return false;
  }
  if (prev.isSelected === next.isSelected) {
    return true;
  }
});
