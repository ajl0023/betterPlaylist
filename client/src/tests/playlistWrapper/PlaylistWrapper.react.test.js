import "@testing-library/jest-dom";
import { cleanup, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import axios from "axios";
import { createMemoryHistory } from "history";
import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { act, create } from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { ReactComponent as CheckIcon } from "../../images/check.svg";
import PlaylistTracks from "../../PlaylistTracks";
import PlaylistWrapper from "../../PlaylistWrapper";
import { render } from "../test-utils/test-utils";
import { reducerObj } from "./mockreducer";
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
jest.mock("axios");
jest.mock("react-router-dom", () => {
  return {
    ...jest.requireActual("react-router-dom"),
    useParams: () => {
      return { id: "testplaylist1" };
    },
    useRouteMatch: () => {
      return { path: "/playlists" };
    },
  };
});
let component;
let renderedComponent;
let componentRoot;
let history;
beforeEach(() => {
  axios.delete.mockResolvedValue({ status: 200, data: {} });
  axios.post.mockResolvedValue({ status: 200, data: {} });
  axios.get.mockImplementation((url) => {
    switch (url) {
      case "/api/playlist/search/?regex=a":
        return Promise.resolve({
          status: 200,
          data: [
            {
              playlistid: "testplaylist1",
              _id: "608a30325d933b44d87ba9bb",
              track: {
                name: "A Song test",
                album: {
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].album,
                },
                artists: [
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].artists,
                ],
              },
            },
          ],
        });
      case "/api/playlist/search/?regex=a&playlist=testplaylist1":
        return Promise.resolve({
          status: 200,
          data: [
            {
              playlistid: "testplaylist1",
              _id: "608a30325d933b44d87ba9bb",
              track: {
                name: "A Song test",
                album: {
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].album,
                },
                artists: [
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].artists,
                ],
              },
            },
          ],
        });
      case "/api/playlist/search/?regex=a&offset=608a30325d933b44d87ba9bb":
        return Promise.resolve({
          status: 200,
          data: [
            {
              playlistid: "testplaylist1",
              _id: "searchresultsscrollid",
              track: {
                name: "search results scroll song name",
                album: {
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].album,
                },
                artists: [
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"].artists,
                ],
              },
            },
          ],
        });
      case "/api/single/playlists/testplaylist1/?offset=1f5bdd0ea054c94c456a5":
        return Promise.resolve({
          status: 200,
          data: {
            offset: "null",
            playlist: { id: "testplaylist1" },
            tracks: [
              {
                playlistid: "testplaylist1",
                track: {
                  ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"],
                  name: "scroll song name",
                },
                _id: "newscrolledTrackid",
              },
            ],
          },
        });
      default:
        return Promise.resolve({
          status: 200,
          data: {
            offset: "1f5bdd0ea054c94c456a5",
            playlist: { ...reducerObj.playlists.byIds.testplaylist1 },
            tracks: [
              {
                playlistid: "testplaylist1",
                _id: "1f5bdd0ea054c94c456a5",
                track: { ...reducerObj.tracks.byIds["1f5bdd0ea054c94c456a5"] },
              },
            ],
          },
        });
    }
  });
  history = createMemoryHistory();
  history.push("/playlists/testplaylist1");
  const store = mockStore(reducerObj);
  const mainComponent = (
    <Router history={history}>
      <Provider store={store}>
        <PlaylistWrapper></PlaylistWrapper>
      </Provider>
    </Router>
  );
  renderedComponent = render(
    <Router history={history}>
      <PlaylistWrapper key={5}></PlaylistWrapper>
    </Router>,
    { initialState: reducerObj }
  );
  component = create(mainComponent);
  componentRoot = component.root;
});
describe("ui changes for singleplaylist component", () => {
  afterEach(cleanup);
  test("correctly handles ui changes when track is selected", async () => {
    const playlistTracks = componentRoot.findAllByType(PlaylistTracks)[0];
    const checkIcon = playlistTracks.findAllByType(CheckIcon)[0];
    let tree;
    await act(async () => {
      await checkIcon.props.onClick();
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("delete modal properly displays", async () => {
    const { getByTestId } = renderedComponent;
    const deleteModal = getByTestId(/delete modal test/);
    const deleteButton = getByTestId(/delete icon/, { exact: false });
    expect(deleteModal.className).toBe("inactive");
    fireEvent(deleteButton, new MouseEvent("click", { bubbles: true }));
  });
  test("track is deleted from the ui", async () => {
    const {
      getByText,
      queryByText,
      getByTestId,
      getByRole,
    } = renderedComponent;
    const checkIcon = getByTestId("1f5bdd0ea054c94c456a5");
    const confirmDelete = getByRole("button", { name: "Delete" });
    userEvent.click(checkIcon);
    await waitFor(() => {
      expect(getByText(/selected 1/i)).toBeInTheDocument();
    });
    expect(queryByText(/song name/i)).toBeInTheDocument();
    await waitFor(() => {
      userEvent.click(confirmDelete);
      expect(queryByText(/song name/i)).not.toBeInTheDocument();
    });
  });
  test("track is added into the ui", async () => {
    const { queryAllByText, getByTestId, getByRole } = renderedComponent;
    const checkIcon = getByTestId("1f5bdd0ea054c94c456a5");
    const addTracksButton = getByRole("button", { name: "Add" });
    act(() => {
      userEvent.click(checkIcon);
    });
    act(() => {
      userEvent.click(getByTestId("addModaltestplaylist1"));
    });
    expect(queryAllByText(/song name/i).length).toBe(1);
    userEvent.click(addTracksButton);
    await waitFor(() => {
      expect(queryAllByText(/song name/i).length).toBe(2);
    });
  });
});
describe("tests for search results", () => {
  test("search results show for all playlists", async () => {
    history.push("/playlists");
    const { queryByText, getByLabelText } = renderedComponent;
    const calls = require("../../spotify-redux/actions/calls");
    calls.testF = jest.fn();
    jest.useFakeTimers();
    const input = getByLabelText("search-bar");
    act(() => {
      fireEvent.change(input, { target: { value: "a" } });
      jest.advanceTimersByTime(300);
    });
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 100);
    await waitFor(async () => {
      expect(queryByText(/A Song test/i)).toBeInTheDocument();
      expect(queryByText(/song name/i)).not.toBeInTheDocument();
    });
  });
  test("search results show for only one playlist", async () => {
    history.push("/playlists/testplaylist1");
    const { queryByText, getByLabelText } = renderedComponent;
    jest.useFakeTimers();
    const input = getByLabelText("search-bar");
    expect(queryByText(/song name/i)).toBeInTheDocument();
    act(() => {
      fireEvent.change(input, { target: { value: "a" } });
      jest.advanceTimersByTime(300);
    });
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 100);
    await waitFor(async () => {
      expect(queryByText(/A Song test/i)).toBeInTheDocument();
      expect(queryByText(/song name/i)).not.toBeInTheDocument();
    });
  });
});
describe("data is fetched when user is scrolled to bottom", () => {
  test("data is fetched on scroll when user is on a single playlist", async () => {
    const { getByText, getByTestId } = renderedComponent;
    const scrollContainer = getByTestId("scroll-container");
    Object.defineProperty(scrollContainer, "clientHeight", {
      configurable: true,
      value: 500,
    });
    Object.defineProperty(scrollContainer, "scrollHeight", {
      configurable: true,
      value: 680,
    });
    fireEvent.scroll(scrollContainer, { target: { scrollTop: 200 } });
    await Promise.resolve();
    expect(getByText(/scroll song name/i)).toBeInTheDocument();
  });
  test("data is fetched on scroll when user is on search results", async () => {
    const {
      getByText,
      getByTestId,
      findByText,
      getByLabelText,
    } = renderedComponent;
    history.push("/playlists");
    const scrollContainer = getByTestId("scroll-container");
    const input = getByLabelText("search-bar");
    jest.useFakeTimers();
    act(() => {
      fireEvent.change(input, { target: { value: "a" } });
      jest.advanceTimersByTime(300);
    });
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 100);
    expect(await findByText(/A Song test/i)).toBeInTheDocument();
    Object.defineProperty(scrollContainer, "clientHeight", {
      configurable: true,
      value: 500,
    });
    Object.defineProperty(scrollContainer, "scrollHeight", {
      configurable: true,
      value: 680,
    });
    fireEvent.scroll(scrollContainer, { target: { scrollTop: 200 } });
    await Promise.resolve();
    expect(getByText(/search results scroll song name/i)).toBeInTheDocument();
  });
});
