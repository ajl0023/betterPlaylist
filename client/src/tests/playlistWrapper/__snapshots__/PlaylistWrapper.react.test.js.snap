// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui changes for singleplaylist component correctly handles ui changes when track is selected 1`] = `
<div
  className="container-container"
  data-testid="scroll-container"
  id="scroll-container"
  onScroll={[Function]}
>
  <div
    className="header-container"
  >
    <div
      className="options-container"
    >
      <div
        className="drop-down-container"
      >
        <div
          className="drop-down-content"
        >
          <div
            className="drop-down-left-container"
          >
            <div
              className="cancel-icon-container"
              onClick={[Function]}
            >
              <span
                className="main-trigger-icon-container"
              >
                <i
                  className="main-trigger-icon"
                />
              </span>
            </div>
            <p
              className="selected-count"
            >
              Selected 
              1
            </p>
          </div>
          <div
            className="main-options-icons"
          >
            <svg
              className="trash-icon"
              data-testid="delete icon"
              onClick={[Function]}
            >
              trash.svg
            </svg>
            <svg
              className="add-icon-container"
              onClick={[Function]}
            >
              add-icon.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      className="inactive"
    >
      <div
        className="inactive"
      >
        <p
          className="title-text"
        >
          Playlists
        </p>
        <div
          className="input-container"
        >
          <input
            aria-label="search-bar"
            className="input-searchbar"
            id=""
            name=""
            onChange={[Function]}
            type="text"
            value=""
          />
          <div
            className="input-content-container"
          >
            <svg
              className="search-icon"
            >
              search.svg
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="wrapper"
  >
    <div>
      <div
        className="inactive"
        data-testid="delete modal test"
      >
        <div
          className="container-mask"
          onClick={[Function]}
        />
        <div
          className="modal-container"
        >
          <div
            className="header-container"
          >
            <div
              className="cancel-icon-container"
              onClick={[Function]}
            >
              <span
                className="main-trigger-icon-container"
              >
                <i
                  className="main-trigger-icon"
                />
              </span>
               
            </div>
          </div>
          <p>
            Are you sure you want to delete these tracks?
          </p>
          <div
            className="button-container"
          >
            <button
              className="cancel-button"
              onClick={[Function]}
            >
              Cancel
            </button>
            <button
              className="delete-button"
              name="delete-button"
              onClick={[Function]}
              role="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
      <div
        className="inactive"
      >
        <div
          className="container-mask"
          onClick={[Function]}
        />
        <div
          className="inactive"
        >
          <div
            className="modal-container"
          >
            <div
              className="header-container"
            >
              <div
                className="cancel-icon-container"
                onClick={[Function]}
              >
                <span
                  className="main-trigger-icon-container"
                >
                  <i
                    className="main-trigger-icon"
                  />
                </span>
                 
              </div>
            </div>
            <p
              className="modal-title"
            >
              Add to playlists
            </p>
            <div
              className="playlist-container"
            >
              <div
                className="playlist-item-container"
              >
                <div
                  className="playlist-item"
                >
                  <div
                    className="check-index-container"
                  >
                    <p
                      className="playlist-count"
                    >
                      1
                    </p>
                    <svg
                      className="check-icon-place"
                      style={
                        Object {
                          "display": "none",
                        }
                      }
                    >
                      checkIconPlaceH.svg
                    </svg>
                    <svg
                      className="check-icon"
                      data-testid="addModal1234"
                      onClick={[Function]}
                      style={
                        Object {
                          "display": "none",
                          "fill": "white",
                        }
                      }
                    >
                      check.svg
                    </svg>
                  </div>
                  <p
                    className="playlist-title"
                  />
                </div>
              </div>
              <div
                className="playlist-item-container"
              >
                <div
                  className="playlist-item"
                >
                  <div
                    className="check-index-container"
                  >
                    <p
                      className="playlist-count"
                    >
                      2
                    </p>
                    <svg
                      className="check-icon-place"
                      style={
                        Object {
                          "display": "none",
                        }
                      }
                    >
                      checkIconPlaceH.svg
                    </svg>
                    <svg
                      className="check-icon"
                      data-testid="addModaltestplaylist1"
                      onClick={[Function]}
                      style={
                        Object {
                          "display": "none",
                          "fill": "white",
                        }
                      }
                    >
                      check.svg
                    </svg>
                  </div>
                  <p
                    className="playlist-title"
                  >
                    My playlist #6
                  </p>
                </div>
              </div>
            </div>
            <div
              className="button-container"
            >
              <button
                className="cancel-button"
                onClick={[Function]}
              >
                Cancel
              </button>
              <button
                className="delete-button"
                onClick={[Function]}
              >
                Add
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="container"
      >
        <div
          className="inactive"
        >
          <div
            className="item-container"
          >
            <a
              className="playlist-container"
              href="undefined/1234"
              onClick={[Function]}
            >
              <div
                className="playlist-content"
              >
                <div
                  className="playlist-image-container"
                >
                  <img
                    alt=""
                    className="playlist-image"
                    src=""
                  />
                </div>
                <div
                  className="description-container"
                >
                  <p
                    className="playlist-title"
                  />
                  <p
                    className="playlist-description"
                  >
                    By Austin Lee
                  </p>
                </div>
              </div>
            </a>
            <a
              className="playlist-container"
              href="undefined/testplaylist1"
              onClick={[Function]}
            >
              <div
                className="playlist-content"
              >
                <div
                  className="playlist-image-container"
                >
                  <img
                    alt=""
                    className="playlist-image"
                    src=""
                  />
                </div>
                <div
                  className="description-container"
                >
                  <p
                    className="playlist-title"
                  >
                    My playlist #6
                  </p>
                  <p
                    className="playlist-description"
                  >
                    By Austin Lee
                  </p>
                </div>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="wrapper"
    >
      <button
        onClick={[Function]}
      >
        Delete All
      </button>
      <button
        onClick={[Function]}
      >
        Add All
      </button>
      <button
        onClick={[Function]}
      >
        get total count
      </button>
      <button />
      <div
        className="header-container"
      >
        <div
          className="album-cover-container"
          style={
            Object {
              "background": "transparent",
            }
          }
        >
          <img
            alt=""
            className="album-cover"
            id="playlistImage"
            src=""
          />
           
          <div
            className="album-mask"
          />
        </div>
        <div
          className="playlist-info-container"
        >
          <h3
            className="playlist-title"
          >
            My playlist #6
          </h3>
          <p
            className="playlist-creator"
          >
            Created By 
            Austin Lee
             
          </p>
           
          <button
            className="update-button"
            onClick={[Function]}
          >
            update
          </button>
        </div>
      </div>
      <div
        className="tracks-container"
      >
        <div
          className="label-container"
        >
          <div
            className="label-check-container"
          >
            <p
              className="label-count"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              #
            </p>
            <svg
              className="check-box"
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                  "fill": "white",
                }
              }
            >
              check.svg
            </svg>
          </div>
          <div
            className="label-title"
          >
            title
          </div>
          <div
            className="label-album"
          >
            album
          </div>
          <div
            className="label-playlist"
          >
            playlist
          </div>
           
        </div>
        <div
          className="item-wrapper"
        >
          <div
            className="item-container"
          >
            <div
              className="check-count-container"
            >
              <div
                className="check-container"
              >
                <svg
                  className="check-icon-place"
                  style={
                    Object {
                      "display": "flex",
                    }
                  }
                >
                  checkIconPlaceH.svg
                </svg>
                <svg
                  className="check-icon"
                  data-testid="6081f5bdd0ea054c94c456a5"
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "flex",
                      "fill": "#1db954",
                    }
                  }
                >
                  check.svg
                </svg>
              </div>
              <div
                className="track-count-container"
              >
                <span
                  className="inactive"
                >
                  1
                </span>
              </div>
            </div>
            <div
              className="item-title-container"
            >
              <div
                className="album-image-container"
              >
                <img
                  alt=""
                  className="album-image"
                  src=""
                />
              </div>
              <div
                className="title-container"
              >
                <div
                  className="track-title-container"
                >
                  <p
                    className="track-title"
                  >
                    Sleepless Nights (feat. Nightly)
                  </p>
                </div>
                <span
                  className="track-artist"
                >
                  <span>
                    <a>
                      <span
                        className="artist-name"
                      >
                        ayokay
                      </span>
                    </a>
                    , 
                  </span>
                  <span>
                    <a>
                      <span
                        className="artist-name"
                      >
                        Nightly
                      </span>
                    </a>
                  </span>
                </span>
              </div>
            </div>
            <div
              className="track-album-container"
            >
              <p
                className="track-album"
              >
                In the Shape of a Dream
              </p>
            </div>
            <div
              className="track-playlist-container"
            >
              <p
                className="track-playlist"
              >
                My playlist #6
              </p>
            </div>
          </div>
          <div
            className="item-container"
          >
            <div
              className="check-count-container"
            >
              <div
                className="check-container"
              >
                <svg
                  className="check-icon-place"
                  style={
                    Object {
                      "display": "flex",
                    }
                  }
                >
                  checkIconPlaceH.svg
                </svg>
                <svg
                  className="check-icon"
                  data-testid="1f5bdd0ea054c94c456a5"
                  onClick={[Function]}
                  style={
                    Object {
                      "display": "none",
                      "fill": "white",
                    }
                  }
                >
                  check.svg
                </svg>
              </div>
              <div
                className="track-count-container"
              >
                <span
                  className="inactive"
                >
                  2
                </span>
              </div>
            </div>
            <div
              className="item-title-container"
            >
              <div
                className="album-image-container"
              >
                <img
                  alt=""
                  className="album-image"
                  src=""
                />
              </div>
              <div
                className="title-container"
              >
                <div
                  className="track-title-container"
                >
                  <p
                    className="track-title"
                  >
                    song name
                  </p>
                </div>
                <span
                  className="track-artist"
                >
                  <span>
                    <a>
                      <span
                        className="artist-name"
                      >
                        ayokay
                      </span>
                    </a>
                    , 
                  </span>
                  <span>
                    <a>
                      <span
                        className="artist-name"
                      >
                        Nightly
                      </span>
                    </a>
                  </span>
                </span>
              </div>
            </div>
            <div
              className="track-album-container"
            >
              <p
                className="track-album"
              >
                album
              </p>
            </div>
            <div
              className="track-playlist-container"
            >
              <p
                className="track-playlist"
              >
                My playlist #6
              </p>
            </div>
          </div>
           
        </div>
      </div>
    </div>
  </div>
   
</div>
`;
